### PR TITLE
Migrated `Checkbox` to Material 3 Colors

### PR DESCRIPTION
**Pull request by QuncCccccc**
Updated checkbox colors so they use the material 3 palette.

Screen Shot 2022-08-29 at 1 44 31 PMScreen Shot 2022-08-29 at 1 46 20 PM

In order to use the `Checkbox` with the new Material 3 defaults, turn on the `useMaterial3` flag in the `ThemeData`:

```dart
  return MaterialApp(
    theme: ThemeData(useMaterial3: true),
    // ...
  );
```

#104446

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Review by HansMuller — COMMENTED** [review]
It seems possible that the implementation of this widget could be made substantially simpler in a future PR. Perhaps the code that computes "effective" colors by adding a state to the widgets actual state, could be deferred. So that the painter only computed those colors when it needed them. The division of labor between the Toggleable classes and this class seems like more overhead than should be needed.

File: packages/flutter/lib/src/material/checkbox.dart [hunk 1]
```diff
@@ -438,40 +429,48 @@ class _CheckboxState extends State<Checkbox> with TickerProviderStateMixin, Togg
...
         ?? widget.hoverColor
         ?? checkboxTheme.overlayColor?.resolve(hoveredStates)
-        ?? themeData.hoverColor;
+        ?? defaults.overlayColor!.resolve(hoveredStates)!;
 
     final Set<MaterialState> activePressedStates = activeStates..add(MaterialState.pressed);
     final Color effectiveActivePressedOverlayColor = widget.overlayColor?.resolve(activePressedStates)
         ?? checkboxTheme.overlayColor?.resolve(activePressedStates)
-        ?? effectiveActiveColor.withAlpha(kRadialReactionAlpha);
+        ?? activeColor?.withAlpha(kRadialReactionAlpha)
```
Comment (guidezpl) [inline packages/flutter/lib/src/material/checkbox.dart:458]: Should `withAlpha...` be part of the defaults class?

**Review by guidezpl — COMMENTED** [review]
A couple of TODOs
- `CheckboxThemeData.side` needs to support `MaterialStateProperty<Side>` instead of `Side`
- Error states

Are you planning on addressing these in this PR or a future one?

**Review by guidezpl — COMMENTED** [review]
One more thing, the `useMaterial3` documentation should be updated

File: packages/flutter/lib/src/material/checkbox.dart (hunk 1 above)
Comment (QuncCccccc) [reply packages/flutter/lib/src/material/checkbox.dart:458]: I added the `.withAlpha(kRadialReactionAlpha)` for the `overlayColor` in the `_CheckboxDefaultsM2` class. This `activeColor` comes from the widget's properties(like activeIconColor, inactiveIconColor, iconColor) and custom CheckboxThemeData, so I just leave it there with the original implementation. It is not related to any default settings.

**Comment by QuncCccccc** [comment]
> A couple of TODOs
> 
> * `CheckboxThemeData.side` needs to support `MaterialStateProperty<Side>` instead of `Side`
> * Error states
> 
> Are you planning on addressing these in this PR or a future one?

I will create a separate PR for the TODOs:) Thanks!

File: packages/flutter/lib/src/material/checkbox.dart [hunk 2]
```diff
@@ -438,40 +429,48 @@ class _CheckboxState extends State<Checkbox> with TickerProviderStateMixin, Togg
...
-    final Color effectiveActiveColor = widget.fillColor?.resolve(activeStates)
-      ?? _widgetFillColor.resolve(activeStates)
-      ?? checkboxTheme.fillColor?.resolve(activeStates)
-      ?? _defaultFillColor.resolve(activeStates);
-    final Color effectiveInactiveColor = widget.fillColor?.resolve(inactiveStates)
-      ?? _widgetFillColor.resolve(inactiveStates)
-      ?? checkboxTheme.fillColor?.resolve(inactiveStates)
-      ?? _defaultFillColor.resolve(inactiveStates);
+    final Color? activeColor = widget.fillColor?.resolve(activeStates)
+        ?? _widgetFillColor.resolve(activeStates)
```
[the remark this answers is not in the document]
Comment (QuncCccccc) [reply packages/flutter/lib/src/material/checkbox.dart]: Done.

File: dev/tools/gen_defaults/lib/checkbox_template.dart [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        return ${componentColor('md.comp.checkbox.unselected.disabled.outline')}.withOpacity(${opacity('md.comp.checkbox.unselected.disabled.container.opacity')});
+      }
+      if (states.contains(MaterialState.selected)) {
+        if (states.contains(MaterialState.focused)) {
+          return ${componentColor('md.comp.checkbox.selected.focus.container')};
+        }
+        if (states.contains(MaterialState.hovered)) {
+          return ${componentColor('md.comp.checkbox.selected.hover.container')};
+        }
+        if (states.contains(MaterialState.pressed)) {
```
[the remark this answers is not in the document]
Comment (QuncCccccc) [reply dev/tools/gen_defaults/lib/checkbox_template.dart]: Fixed. Thanks for catching this!

File: packages/flutter/lib/src/material/checkbox.dart (hunk 1 above)
Comment (guidezpl) [reply packages/flutter/lib/src/material/checkbox.dart]: Gotcha

File: packages/flutter/lib/src/material/checkbox.dart [hunk 4]
```diff
@@ -683,3 +682,172 @@ class _CheckboxPainter extends ToggleablePainter {
...
+  MaterialStateProperty<Color> get fillColor {
+    return MaterialStateProperty.resolveWith((Set<MaterialState> states) {
+      if (states.contains(MaterialState.disabled)) {
+        if (states.contains(MaterialState.selected)) {
+          return _colors.onSurface.withOpacity(0.38);
+        }
+        return _colors.onSurface.withOpacity(0.38);
+      }
+      if (states.contains(MaterialState.selected)) {
+        if (states.contains(MaterialState.pressed)) {
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/checkbox.dart:766]: TODO (future PR): update the code generator to collapse code like this.

File: packages/flutter/lib/src/material/checkbox.dart [hunk 5]
```diff
@@ -683,3 +682,172 @@ class _CheckboxPainter extends ToggleablePainter {
...
+      }
+      if (states.contains(MaterialState.focused)) {
+        return _colors.onSurface;
+      }
+      return _colors.onSurface;
+    });
+  }
+
+  @override
+  MaterialStateProperty<Color?> get checkColor {
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/checkbox.dart:791]: It looks like we should now be returning `MaterialStateProperty<Color>` in methods like this one.

**Review by HansMuller — APPROVED** [review]
LGTM

**Review by guidezpl — APPROVED** [review]
Edited the description to not close the issue, pending the remaining work.

**Pull request merged**